afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+#![deny(missing_docs)]
+
+use libc;
+use std::{
+    io,
+    ptr,
+    mem::{self, MaybeUninit},
+    ffi::CStr};
+use crate::linux;
+
+/// Represents a link-local address.
+/// At this time, it's not particularly useful.
+pub struct Addr {
+    inner: libc::sockaddr_storage,
+    len: libc::socklen_t,
+}
+
+/// Represents an unbound `AF_PACKET` socket.  At this phase of a socket's lifecycle, it can be
+/// configured.
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
+        // syscall return value and returns an Err if the sycall fails. It does not interact with
+        // Rust-owned memory, so it's safe as per the borrow checker's rules.
+        let fd = unsafe {
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#socket
+            // man 7 packet
+            let fd = libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL.to_be());
+            if fd < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            fd
+        };
+        Ok(Self { fd })
+    }
+
+    /// Binds the socket to a network interface. This function consumes the `Socket` instance, as
+    /// no more configuration options may be safely changed.
+    pub fn bind(self, iface: impl AsRef<CStr>) -> io::Result<BoundSocket> {
```

Comment:
The interface to `bind` takes a `sockaddr` as a descriptor to where to bind the given `Socket`. `sockaddr` is essentially a struct with IP and port, so why does our `Socket#bind` take a C string? Is there a reason we can't bind the type more strictly?

Reply:
Beej's guide doesn't really cover `AF_PACKET` sockets, since it's about stuff that's common to all Unix OSes. Again, noting the [man page](http://man7.org/linux/man-pages/man7/packet.7.html), it takes a `sockaddr_ll` struct, which represents the network interface. That's why I'm taking a `&CStr` -- it's supposed to represent the name of the network interface the caller wants to bind this socket to.

We could define a `NetworkInterface` struct and pass that in instead of a `&CStr`.

Reply:
Yeah a `NetworkInterface` would be preferable.

Reply:
What sort of data should this `NetworkInterface` expose?
Aside, I found [this](http://man7.org/linux/man-pages/man3/if_nametoindex.3.html), which could simplify the `ifreq` dance.

Reply:
Discussed offline, the network interface here is the _NIC_ name, which doesn't have a well-defined structure at compile-time.